Add drop_privileges() for Linux

viraptor opened the pull request:
Implement an aggressive version of drop_privileges(). Additionally add
similar initialization function for threads drop_worker_privileges().

This version is similar to Solaris one and prohibits memcached from
making any not approved syscalls. Current list narrows down the allowed
calls to socket sends/recvs, accept, epoll handling, futex (and
dependencies - mmap), getrusage (for stats), and signal / exit
handling.

Any incorrect behaviour will result in EACCES returned. This should be
restricted further to KILL in the future (after more testing).

The feature is only tested for i386 and x86_64. It depends on bpf
filters and seccomp enabled in the kernel. It also requires libsecomp
for abstraction to seccomp filters. All are available since Linux 3.5.

Seccomp filtering can be enabled at compile time with --enable-seccomp.

In case of local customisations which require more rights, memcached
allows disabling drop_privileges() with "-o no_drop_privileges" at
startup.

Tests have to run with "-o relaxed_privileges", since they require
disk access after the tests complete. This adds a few allowed syscalls,
but does not disable the protection system completely.

jakesyl commented:
+1 on merge

viraptor commented:
This pr needs a refresh to work with me changes, but I wanted to check if anyone is interested in it first.
If there's a chance it will be looked at, I'm happy to refresh it.

dormando commented:
I can look, but it's very difficult for me to test it. It'd help if we can find some other people to do so, or figure out how I can get one of my cloud buildbots to do it. Any thoughts there?

viraptor commented:
@dormando Are you thinking of automated negative tests? As in verification that the blocking actually happens? The only thing I can think of right now is adding a command that's only available in the test build and only exists to trigger things that shouldn't happen. For example `DO_BAD_THING 1` tries to `system("any_command")`, `DO_BAD_THING 2` tries to bind a new listening socket, etc.

As for positive testing, the whole test suite should work just fine with slightly extended rights. They do need extended privileges as mentioned in the commit - `-o relaxed_privileges` allows reporting all the test results properly, but doesn't actually kill the protection system.

I can't find any configuration for the buildbot in the repository. If there's anything I can help there, I'm happy to have a look.

Let me know if that's what you had in mind.

dormando commented:
Need at least positive tests. Maybe this isn't as hard as I was thinking - but verifying that the code is getting some exercize on my builders. The pattern for those kind of tests is looking for a magic environment variable that then tries to do something... but that might end up in someone's production, so the test command would have to be carefully chosen to prevent a security problem for people who don't have the privdrop enabled.

Buildbot stuff isn't in the repo sadly, it's tied to my cloud accounts. You can browse it a bit here: http://build.memcached.org:8010/ - the master is always live, and I fire up the builders when I'm working on a release.

There's also this that came in recently: https://github.com/memcached/memcached/pull/203 - not sure if there's any common ground or improvements to be had there.

viraptor commented:
@dormando I added a test that tries to run a command and open a socket in the worker thread - both operations are stopped by seccomp.

The command itself is only built in memcached-debug. There's a MEMCACHED_DEBUG define flag available for it now. It also enables more allowed calls automatically so the performance profile can be saved.

The branch is now up to date with the current master. This includes the lru worker and everything that runs during `make test`. I also added some details in HACKING about how to debug issues of seccomp breaking new code.

dormando commented:
(late, again :\)

Are you missing a file from the commit? It doesn't build due to missing -lseccomp from LIBS flags. I see it being added in Makefile.am under LDCONFIG but that's not propagating. How were you building this?

(I can fix it; but generally think it's better to ask when I see something odd in this way).

Also, if you have a minute; can you add the the extra outputs to doc/protocol.txt and rebase against `next`?

dormando commented:
one more poke. since it wasn't building I hadn't returned to this.

viraptor commented:
Missed your comment, thanks for the poke! I'll have a look at this in the next day or two.

dormando commented:
raaad, thanks!

viraptor commented:
I uploaded the code rebased on current `next` branch. (https://github.com/memcached/memcached/pull/275) It required one extra syscall and has the extra command documented.

I'm happy to debug the build issue, but I can't reproduce it :( On a clean checkout `./autogen.sh ; ./configure --enable-seccomp ; make ; make test` just works for me. What's the error you're seeing?

dormando commented:
Sorry, again :( I immediately got sick after you rebased things.

Here's the error during compilation: https://gist.github.com/dormando/8069d6bb920eea03dc0db199ea61c570

It configures fine. As I said before I don't see -lseccomp being added anywhere, so it can't link toward the end there.

dormando commented:
err nevermind, I'm blind.. I see the -lseccomp. not obvious why it's failing though.

viraptor commented:
Ohh... interesting. What system is that on? I've tried on Arch and Debian without issues.

dormando commented:
ubuntu 16.04

dormando commented:
Any ideas (or something I should look for?) I'm probably cutting a release soon.

jakesyl commented:
Huh, that's wierd.  Works fine on my Ubuntu 16.04 version.

viraptor commented:
Sorry, I've been away the last 2 weeks. Will check it tonight.

viraptor commented:
I did get the error by staring from a clean `ubuntu:16.04` docker image:
```
apt install build-essential libevent-dev libseccomp-dev
./autogen.sh
./configure --enable-seccomp
make
```
Getting same error about missing `seccomp_rule_add` in that case. I'll investigate some more soon.

viraptor commented:
Fixed in the next branch PR (https://github.com/memcached/memcached/pull/275)

dormando commented:
merged fia the other PR
